ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ReadWriteServiceBridge.scala

```diff
@@ -18,37 +17,227 @@ import com.daml.ledger.configuration.{
+import com.daml.lf.transaction.{BlindingInfo, CommittedTransaction, GlobalKey, SubmittedTransaction}
 import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.apiserver.execution.MissingContracts
+import com.daml.platform.store.appendonlydao.events._
 import com.daml.telemetry.TelemetryContext
 import com.google.common.primitives.Longs
 import com.google.rpc.code.Code
 import com.google.rpc.status.Status
 
+import java.util.UUID
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+import java.util.concurrent.{CompletableFuture, CompletionStage}
+import scala.collection.Searching
+import scala.concurrent.{ExecutionContextExecutor, Future}
+import scala.util.{Failure, Success, Try}
+
 case class ReadWriteServiceBridge(
     participantId: Ref.ParticipantId,
     ledgerId: LedgerId,
     maxDedupSeconds: Int,
     submissionBufferSize: Int,
+    contractStore: AtomicReference[Option[ContractStore]],
 )(implicit mat: Materializer, loggingContext: LoggingContext)
     extends ReadService
     with WriteService
     with AutoCloseable {
+  private val offsetIdx = new AtomicLong(0L)
+
   import ReadWriteServiceBridge._
+  private implicit val ec: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
 
   private[this] val logger = ContextualizedLogger.get(getClass)
 
+  private val (conflictCheckingQueue, source) = Source
+    .queue[Submission.Transaction](128)
+    .mapAsyncUnordered(64) { tx =>
+      val enqueuedAt = contractStore
+        .get()
+        .getOrElse(throw new RuntimeException("ContractStore not there yet"))
+        .cacheOffset()
+      parallelConflictCheck(tx).map(enqueuedAt -> _)
+    }
+    .statefulMapConcat(sequence)
+    .preMaterialize()
+
+  private def sequence: () => ((Offset, Either[SoxRejection, Transaction])) => Iterable[
+    (Offset, Update)
+  ] = () => {
+    @volatile var sequencerQueue = Vector.empty[(Offset, Submission.Transaction)]
+
+    {
+      case (_, Left(rejection)) =>
+        val newIndex = offsetIdx.getAndIncrement()
+        val newOffset = toOffset(newIndex)
+        Iterable(newOffset -> toRejection(rejection))
+      case (enqueuedAt, Right(transaction)) =>
+        val newIndex = offsetIdx.getAndIncrement()
+        val newOffset = toOffset(newIndex)
+        checkSequential(enqueuedAt, transaction, sequencerQueue) match {
+          case Left(rejection) =>
+            Iterable(
+              newOffset -> toRejection(rejection)
+            )
+          case Right(acceptedTx) =>
+            sequencerQueue = sequencerQueue :+ (newOffset -> transaction)
+            Iterable(newOffset -> successMapper(acceptedTx, newIndex, participantId))
+        }
+    }
+  }
+
+  private def conflictCheckSlice(
+      delta: Vector[(Offset, Submission.Transaction)],
+      transaction: Submission.Transaction,
+  ): Either[SoxRejection, Submission.Transaction] = {
+    val (cKeys, _, archives) = collectDelta(delta)
+    transaction.transaction
```

Comment:
I strongly recommend that you follow the approach used in the kv committer where you don’t traverse the transaction and instead rely on `contractKeyInputs` and friends and only compar that.